<a href="https://colab.research.google.com/github/Yanhuijun1911/PythonData/blob/main/15_1_Encoding_and_Dummy_Coding.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Encoding Data  
When we encode data, we remodel a numeric column in a dataframe to be on a standard scale (0 or 1).   
For example if we had a column of BMI scores, we could encode that column so that all scores greater than or equal to 25 were recoded to the value 1 (bad) and all scores less than 25 were recoded to 0 (good).  

For example:  

` def encode_bmi(df):`       
> `if df['bmi'] >= 25:`  
> >  `return 1`   

>`else`:  
> >`return 0`  

`df["bmi"] = df.apply(encode_bmi, axis=1)`    

### Challenge 1 - prepare dataset for encoding 
---
1. Read Covid vaccination data from the `by_country` sheet in the Excel file at this link : https://github.com/lilaceri/Working-with-data-/blob/342abab10d93c4bf23b5c55a50f189f12a137c5f/Data%20Sets%20for%20code%20divisio/Covid%20Vaccination%20Data.xlsx?raw=true
2. Find out which columns have missing values
3. Remove all rows with missing data in the total_vaccination column  
4. Remove all rows with missing data in the daily_vaccinations_per_million 
4. find the median daily_vaccinations_per_million, storing this in a variable for use later     


**Test output**:  
1. dataframe is saved in a variable
2. 
```
RangeIndex: 14994 entries, 0 to 14993
Data columns (total 15 columns):
    Column                               Non-Null Count  Dtype         
                                
 0   country                              14994 non-null  object        
 1   iso_code                             14994 non-null  object        
 2   date                                 14994 non-null  datetime64[ns]
 3   total_vaccinations                   9011 non-null   float64       
 4   people_vaccinated                    8370 non-null   float64       
 5   people_fully_vaccinated              6158 non-null   float64       
 6   daily_vaccinations_raw               7575 non-null   float64       
 7   daily_vaccinations                   14796 non-null  float64       
 8   total_vaccinations_per_hundred       9011 non-null   float64       
 9   people_vaccinated_per_hundred        8370 non-null   float64       
 10  people_fully_vaccinated_per_hundred  6158 non-null   float64       
 11  daily_vaccinations_per_million       14796 non-null  float64       
 12  vaccines                             14994 non-null  object        
 13  source_name                          14994 non-null  object        
 14  source_website                       14994 non-null  object        
dtypes: datetime64[ns](1), float64(9), object(5)
memory usage: 1.7+ MB
```
3. 9011 rows × 15 columns  
4. 8815 rows * 15 columns 
5. 6.65 



In [2]:
import pandas as pd
url = "https://github.com/lilaceri/Working-with-data-/blob/342abab10d93c4bf23b5c55a50f189f12a137c5f/Data%20Sets%20for%20code%20divisio/Covid%20Vaccination%20Data.xlsx?raw=true"

# Task 1 
df = pd.read_excel(url,"by_country")


# Task 2
df.isnull()

df = df.dropna(subset = ['total_vaccinations'])
df_new = df.dropna(subset = ['daily_vaccinations_per_million'])



# Task 3
#df =df.dropna(subset = ['total_vaccinations','daily_vaccinations_per_million'], how = 'any')
print(df_new.info())

df.isnull().sum()
# Task 4

median = df_new['daily_vaccinations_per_million'].median()
print(median)



<class 'pandas.core.frame.DataFrame'>
Int64Index: 8816 entries, 6 to 14993
Data columns (total 15 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   country                              8816 non-null   object        
 1   iso_code                             8816 non-null   object        
 2   date                                 8816 non-null   datetime64[ns]
 3   total_vaccinations                   8816 non-null   float64       
 4   people_vaccinated                    8126 non-null   float64       
 5   people_fully_vaccinated              6143 non-null   float64       
 6   daily_vaccinations_raw               7575 non-null   float64       
 7   daily_vaccinations                   8816 non-null   float64       
 8   total_vaccinations_per_hundred       8816 non-null   float64       
 9   people_vaccinated_per_hundred        8126 non-null   float64       
 10  people_full

### Challenge 2 - encode daily vaccinations 
---

Write a function to encode daily vaccinations per million, where values greater than or equal to median = 1 and values less than median = 0 

**Test output**: 

using describe()
```
count    8816.000000
mean        0.991493
std         0.091847
min         0.000000
25%         1.000000
50%         1.000000
75%         1.000000
max         1.000000
Name: daily_vaccinations_per_million, dtype: float64
```

In [3]:
# Task
def encode_daily(df, **kwds):
  median = kwds['median']
  key = kwds ['key']
  if df[key] >= median:
    return 1
  else:
    return 0 

df_new['encode'] = df_new.apply(encode_daily, axis = 1, key = 'daily_vaccinations_per_million', median = median)
df_new.daily_vaccinations_per_million.describe()




/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  # Remove the CWD from sys.path while we load stuff.


count      8816.000000
mean       3352.155626
std        4980.979866
min           0.000000
25%         658.000000
50%        1915.500000
75%        4388.000000
max      118759.000000
Name: daily_vaccinations_per_million, dtype: float64

### Challenge 3 - Encoding total vaccinations   
---
The United Kingdom has been praised for its fast vaccine rollout. 
1. Find the minimum total vaccinations for the United Kingdom 
2. Write a function to encode total_vaccinations column so that all values less than the UK's min are 0 and all values greater than or equal to the UK's min are coded as 1 
3. Display the unique countries for which total vaccinated is at the same rate or more than the UK

**Test output**:

1. 1402432.0
2. `df['total_vaccinations'].describe()` should output:
```
count    9011.00000
mean        0.29808
std         0.45744
min         0.00000
25%         0.00000
50%         0.00000
75%         1.00000
max         1.00000
Name: total_vaccinations, dtype: float64
```
3. 
```
array(['Argentina', 'Australia', 'Austria', 'Azerbaijan', 'Bangladesh',
       'Belgium', 'Brazil', 'Cambodia', 'Canada', 'Chile', 'China',
       'Colombia', 'Czechia', 'Denmark', 'Dominican Republic', 'England',
       'Finland', 'France', 'Germany', 'Greece', 'Hong Kong', 'Hungary',
       'India', 'Indonesia', 'Ireland', 'Israel', 'Italy', 'Japan',
       'Kazakhstan', 'Malaysia', 'Mexico', 'Morocco', 'Nepal',
       'Netherlands', 'Norway', 'Pakistan', 'Peru', 'Philippines',
       'Poland', 'Portugal', 'Qatar', 'Romania', 'Russia', 'Saudi Arabia',
       'Scotland', 'Serbia', 'Singapore', 'Slovakia', 'South Korea',
       'Spain', 'Sweden', 'Switzerland', 'Thailand', 'Turkey',
       'United Arab Emirates', 'United Kingdom', 'United States',
       'Uruguay', 'Wales'], dtype=object)
```




In [4]:
# Task 1
min_UK = df[df['country']== 'United Kingdom']['total_vaccinations'].min()

#min_UK = df_UK['total_vaccinations'].min()
print(min_UK)

# Task 2

df['total_vaccinations'].describe()



1402432.0


count    9.011000e+03
mean     4.964584e+06
std      2.033452e+07
min      0.000000e+00
25%      6.000000e+04
50%      4.209400e+05
75%      1.932222e+06
max      2.753380e+08
Name: total_vaccinations, dtype: float64

In [5]:
def encode_total_vaccinations (df, **kws):
  #min_UK = df[df['country']== 'United Kingdom']['total_vaccinations'].min()
  #key = kws['key']
  if df['total_vaccinations'] >= min_UK:
    return 1
  else:
    return 0

df['Encode_total_vaccinations']= df.apply(encode_total_vaccinations, axis = 1,)

df[df['Encode_total_vaccinations']==1]['country'].unique()
#df_challenge3['country'].unique()





array(['Argentina', 'Australia', 'Austria', 'Azerbaijan', 'Bangladesh',
       'Belgium', 'Brazil', 'Cambodia', 'Canada', 'Chile', 'China',
       'Colombia', 'Czechia', 'Denmark', 'Dominican Republic', 'England',
       'Finland', 'France', 'Germany', 'Greece', 'Hong Kong', 'Hungary',
       'India', 'Indonesia', 'Ireland', 'Israel', 'Italy', 'Japan',
       'Kazakhstan', 'Malaysia', 'Mexico', 'Morocco', 'Nepal',
       'Netherlands', 'Norway', 'Pakistan', 'Peru', 'Philippines',
       'Poland', 'Portugal', 'Qatar', 'Romania', 'Russia', 'Saudi Arabia',
       'Scotland', 'Serbia', 'Singapore', 'Slovakia', 'South Korea',
       'Spain', 'Sweden', 'Switzerland', 'Thailand', 'Turkey',
       'United Arab Emirates', 'United Kingdom', 'United States',
       'Uruguay', 'Wales'], dtype=object)

### Challenge 4 - create new series of total vaccinations for each manufacturer
---

To create a new column in your dataframe:

`df['new_column'] = ...`

For example:

* to duplicate an existing column
  * `df['new_column'] = df['old_column']`
* to add two columns together 
  * `df['new_column'] = df['column1'] + df['column2']`
* to make a percentages column 
  * `df['new_column'] = (df['column1']/df['column1].sum()) * 100`

  
1. read data from 'by_manufacturer' sheet from Covid data 
2. find the sum of total vaccinations for each manufacturer
3. create a new column that has the total vaccinations as a percentage of the overall sum of total vaccinations 
4. find the median percentage 
5. create a new column called 'encoded_percentages' which duplicates the percentages column
6. encode the encoded_percentages column so that any values greater than or equal to the median percentage = 1 and any lesser than = 0 


**Test output**:

1.
2. 
```
vaccine
Johnson&Johnson        264839828
Moderna               5548036383
Oxford/AstraZeneca     539433203
Pfizer/BioNTech       8690461304
Sinovac                604660293
Name: total_vaccinations, dtype: int64
```
3. 
```
	location	date	vaccine	total_vaccinations	percentages
0	Chile	2020-12-24	Pfizer/BioNTech	420	0.000003
1	Chile	2020-12-25	Pfizer/BioNTech	5198	0.000033
2	Chile	2020-12-26	Pfizer/BioNTech	8338	0.000053
3	Chile	2020-12-27	Pfizer/BioNTech	8649	0.000055
4	Chile	2020-12-28	Pfizer/BioNTech	8649	0.000055
...	...	...	...	...	...
3291	United States	2021-05-01	Moderna	105947940	0.677095
3292	United States	2021-05-01	Pfizer/BioNTech	129013657	0.824504
3293	United States	2021-05-02	Johnson&Johnson	8374395	0.053519
3294	United States	2021-05-02	Moderna	106780082	0.682413
3295	United States	2021-05-02	Pfizer/BioNTech	130252779	0.832423
3296 rows × 5 columns
```
4. 0.0011110194374896931
5. 
```
location	date	vaccine	total_vaccinations	percentage_of_total	encoded_percentages
0	Chile	2020-12-24	Pfizer/BioNTech	420	0.000003	0.000003
1	Chile	2020-12-25	Pfizer/BioNTech	5198	0.000033	0.000033
2	Chile	2020-12-26	Pfizer/BioNTech	8338	0.000053	0.000053
3	Chile	2020-12-27	Pfizer/BioNTech	8649	0.000055	0.000055
4	Chile	2020-12-28	Pfizer/BioNTech	8649	0.000055	0.000055
...	...	...	...	...	...	...
3291	United States	2021-05-01	Moderna	105947940	0.677095	0.677095
3292	United States	2021-05-01	Pfizer/BioNTech	129013657	0.824504	0.824504
3293	United States	2021-05-02	Johnson&Johnson	8374395	0.053519	0.053519
3294	United States	2021-05-02	Moderna	106780082	0.682413	0.682413
3295	United States	2021-05-02	Pfizer/BioNTech	130252779	0.832423	0.832423
3296 rows × 6 columns
```
6. 
```
	location	date	vaccine	total_vaccinations	percentages	encode	encoded
0	Chile	2020-12-24	Pfizer/BioNTech	420	0.000003	0.000003	0
1	Chile	2020-12-25	Pfizer/BioNTech	5198	0.000033	0.000033	0
2	Chile	2020-12-26	Pfizer/BioNTech	8338	0.000053	0.000053	0
3	Chile	2020-12-27	Pfizer/BioNTech	8649	0.000055	0.000055	0
4	Chile	2020-12-28	Pfizer/BioNTech	8649	0.000055	0.000055	0
...	...	...	...	...	...	...	...
3291	United States	2021-05-01	Moderna	105947940	0.677095	0.677095	1
3292	United States	2021-05-01	Pfizer/BioNTech	129013657	0.824504	0.824504	1
3293	United States	2021-05-02	Johnson&Johnson	8374395	0.053519	0.053519	1
3294	United States	2021-05-02	Moderna	106780082	0.682413	0.682413	1
3295	United States	2021-05-02	Pfizer/BioNTech	130252779	0.832423	0.832423	1
3296 rows × 7 columns
```



In [6]:
#1.read data from 'by_manufacturer' sheet from Covid data
#2.find the sum of total vaccinations for each manufacturer
#3.create a new column that has the total vaccinations as a percentage of the overall sum of total vaccinations
#4.find the median percentage
#5.create a new column called 'encoded_percentages' which duplicates the percentages column
#6.encode the encoded_percentages column so that any values greater than or equal to the median percentage = 1 and any lesser than = 0

In [7]:
#1/2
import pandas as pd

df_manufacturer = pd.read_excel(url,'by_manufacturer')
df_manufacturer.info()
print(df_manufacturer['vaccine'].unique())
df_Pfizer = df_manufacturer[df_manufacturer['vaccine']=='Pfizer/BioNTech']
df_Sinovac = df_manufacturer[df_manufacturer['vaccine']=='Sinovac']
df_Moderna = df_manufacturer[df_manufacturer['vaccine']=='Moderna']
df_Oxford = df_manufacturer[df_manufacturer['vaccine']=='Oxford/AstraZeneca']
df_Johnson = df_manufacturer[df_manufacturer['vaccine']=='Johnson&Johnson']

sum_Pfizer = df_Pfizer['total_vaccinations'].sum()
sum_Sinovac = df_Sinovac['total_vaccinations'].sum()
sum_Moderna = df_Moderna['total_vaccinations'].sum()
sum_Oxford = df_Oxford['total_vaccinations'].sum()
sum_Johnson = df_Johnson['total_vaccinations'].sum()


print(sum_Pfizer)
print(sum_Sinovac)
print(sum_Moderna)
print(sum_Oxford)
print(sum_Johnson)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3296 entries, 0 to 3295
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   location            3296 non-null   object        
 1   date                3296 non-null   datetime64[ns]
 2   vaccine             3296 non-null   object        
 3   total_vaccinations  3296 non-null   int64         
dtypes: datetime64[ns](1), int64(1), object(2)
memory usage: 103.1+ KB
['Pfizer/BioNTech' 'Sinovac' 'Moderna' 'Oxford/AstraZeneca'
 'Johnson&Johnson']
8690461304
604660293
5548036383
539433203
264839828


In [8]:
#3
sum = sum_Pfizer + sum_Johnson + sum_Moderna + sum_Oxford + sum_Sinovac
df_manufacturer['percentages'] = df_manufacturer['total_vaccinations']/sum
print(df_manufacturer)

           location       date  ... total_vaccinations   percentages
0             Chile 2020-12-24  ...                420  2.684147e-08
1             Chile 2020-12-25  ...               5198  3.321951e-07
2             Chile 2020-12-26  ...               8338  5.328670e-07
3             Chile 2020-12-27  ...               8649  5.527425e-07
4             Chile 2020-12-28  ...               8649  5.527425e-07
...             ...        ...  ...                ...           ...
3291  United States 2021-05-01  ...          105947940  6.770948e-03
3292  United States 2021-05-01  ...          129013657  8.245038e-03
3293  United States 2021-05-02  ...            8374395  5.351930e-04
3294  United States 2021-05-02  ...          106780082  6.824129e-03
3295  United States 2021-05-02  ...          130252779  8.324228e-03

[3296 rows x 5 columns]


In [9]:
#4
median = df_manufacturer['percentages'].median()
median

1.1110194374896931e-05

In [10]:
#5/6
def encoded_percentages(df,**kwds):
  key = kwds['key']
  if df[key] >= median:
    return 1
  else:
    return 0

df_manufacturer['encoded_percentages'] = df_manufacturer.apply(encoded_percentages, axis = 1, key = 'percentages')
print(df_manufacturer)

           location       date  ...   percentages  encoded_percentages
0             Chile 2020-12-24  ...  2.684147e-08                    0
1             Chile 2020-12-25  ...  3.321951e-07                    0
2             Chile 2020-12-26  ...  5.328670e-07                    0
3             Chile 2020-12-27  ...  5.527425e-07                    0
4             Chile 2020-12-28  ...  5.527425e-07                    0
...             ...        ...  ...           ...                  ...
3291  United States 2021-05-01  ...  6.770948e-03                    1
3292  United States 2021-05-01  ...  8.245038e-03                    1
3293  United States 2021-05-02  ...  5.351930e-04                    1
3294  United States 2021-05-02  ...  6.824129e-03                    1
3295  United States 2021-05-02  ...  8.324228e-03                    1

[3296 rows x 6 columns]


### Exercise 8 - encode daily vaccinations 

1. find the median daily vaccinations per 1 million 
2. write a function to encode daily vaccinations per 1 million, where values greater than or equal to median = 1 and values less than median = 0 

Output: 

1. 1915.5
2. 
```
0        0
6        0
22       0
44       0
59       0
        ..
14989    0
14990    0
14991    0
14992    0
14993    0
Name: daily_vaccinations_per_million, Length: 9011, dtype: int64
```

In [15]:
median_daily_vac = df['daily_vaccinations_per_million'].median()
print(median_daily_vac)

def encode_daily_vac(df,**kwds):
  key = kwds['key']
  if df[key] >= median_daily_vac:
    return 1
  else:
    return 0

df['encode_daily_vac'] = df.apply(encode_daily_vac, axis = 1, key = 'daily_vaccinations_per_million')

print(df)

1915.5
           country iso_code  ... Encode_total_vaccinations  encode_daily_vac
0      Afghanistan      AFG  ...                         0                 0
6      Afghanistan      AFG  ...                         0                 0
22     Afghanistan      AFG  ...                         0                 0
44     Afghanistan      AFG  ...                         0                 0
59     Afghanistan      AFG  ...                         0                 0
...            ...      ...  ...                       ...               ...
14989     Zimbabwe      ZWE  ...                         0                 0
14990     Zimbabwe      ZWE  ...                         0                 0
14991     Zimbabwe      ZWE  ...                         0                 0
14992     Zimbabwe      ZWE  ...                         0                 0
14993     Zimbabwe      ZWE  ...                         0                 0

[9011 rows x 17 columns]


### Exercise 9 - Encoding vaccinations per hundred  
---
The United Kingdom has been praised for its fast vaccine rollout. 
1. find the minimum total vaccinations for the United Kingdom 
2. save this value in a variable rounded down to an integer
3. write a function to encode total_vaccinations column so that all values less than the UK's min are 0 and all values greater than or equal to the UK's min are coded as 1 
4. display the countries which total vaccinated is at the same rate or more than the UK

Output:

1. 1402432.0
2. 1402432
3. `df['people_vaccinated_per_hundred']` should output:
```
0        0
6        0
22       0
44       0
59       0
        ..
14989    0
14990    0
14991    0
14992    0
14993    0
Name: total_vaccinations, Length: 9011, dtype: int64
```
4. 
```
array(['Argentina', 'Australia', 'Austria', 'Azerbaijan', 'Bangladesh',
       'Belgium', 'Brazil', 'Cambodia', 'Canada', 'Chile', 'China',
       'Colombia', 'Czechia', 'Denmark', 'Dominican Republic', 'England',
       'Finland', 'France', 'Germany', 'Greece', 'Hong Kong', 'Hungary',
       'India', 'Indonesia', 'Ireland', 'Israel', 'Italy', 'Japan',
       'Kazakhstan', 'Malaysia', 'Mexico', 'Morocco', 'Nepal',
       'Netherlands', 'Norway', 'Pakistan', 'Peru', 'Philippines',
       'Poland', 'Portugal', 'Qatar', 'Romania', 'Russia', 'Saudi Arabia',
       'Scotland', 'Serbia', 'Singapore', 'Slovakia', 'South Korea',
       'Spain', 'Sweden', 'Switzerland', 'Thailand', 'Turkey',
       'United Arab Emirates', 'United Kingdom', 'United States',
       'Uruguay', 'Wales'], dtype=object)
```




### Exercise 10 - create new series of total vaccinations percentages
---

To create a new column in your dataframe:

`df['new_column'] = ...`

For example:

* to duplicate an existing column
  * `df['new_column'] = df['old_column']`
* to add two columns together 
  * `df['new_column'] = df['column1'] + df['column2']`
* to make a percentages column 
  * `df['new_column'] = (df['column1']/df['column1].sum()) * 100`  
  


1. read data from 'by_manufacturer' sheet from Covid data 
2. find the sum of total vaccinations for each manufacturer
3. create a new column that has the total vaccinations as a percentage of the overall sum of total vaccinations 
4. find the median percentage 
5. create a new column called 'encoded_percentages' which duplicates the percentages column
6. encode the encoded_percentages column so that any values greater than or equal to the median percentage = 1 and any lesser than = 0 


Output:

1.
2. 
```
vaccine
Johnson&Johnson        264839828
Moderna               5548036383
Oxford/AstraZeneca     539433203
Pfizer/BioNTech       8690461304
Sinovac                604660293
Name: total_vaccinations, dtype: int64
```
3. 
```
	location	date	vaccine	total_vaccinations	percentages
0	Chile	2020-12-24	Pfizer/BioNTech	420	0.000003
1	Chile	2020-12-25	Pfizer/BioNTech	5198	0.000033
2	Chile	2020-12-26	Pfizer/BioNTech	8338	0.000053
3	Chile	2020-12-27	Pfizer/BioNTech	8649	0.000055
4	Chile	2020-12-28	Pfizer/BioNTech	8649	0.000055
...	...	...	...	...	...
3291	United States	2021-05-01	Moderna	105947940	0.677095
3292	United States	2021-05-01	Pfizer/BioNTech	129013657	0.824504
3293	United States	2021-05-02	Johnson&Johnson	8374395	0.053519
3294	United States	2021-05-02	Moderna	106780082	0.682413
3295	United States	2021-05-02	Pfizer/BioNTech	130252779	0.832423
3296 rows × 5 columns
```
4. 0.0011110194374896931
5. 
6. 
```
	location	date	vaccine	total_vaccinations	percentages	encode	encoded
0	Chile	2020-12-24	Pfizer/BioNTech	420	0.000003	0.000003	0
1	Chile	2020-12-25	Pfizer/BioNTech	5198	0.000033	0.000033	0
2	Chile	2020-12-26	Pfizer/BioNTech	8338	0.000053	0.000053	0
3	Chile	2020-12-27	Pfizer/BioNTech	8649	0.000055	0.000055	0
4	Chile	2020-12-28	Pfizer/BioNTech	8649	0.000055	0.000055	0
...	...	...	...	...	...	...	...
3291	United States	2021-05-01	Moderna	105947940	0.677095	0.677095	1
3292	United States	2021-05-01	Pfizer/BioNTech	129013657	0.824504	0.824504	1
3293	United States	2021-05-02	Johnson&Johnson	8374395	0.053519	0.053519	1
3294	United States	2021-05-02	Moderna	106780082	0.682413	0.682413	1
3295	United States	2021-05-02	Pfizer/BioNTech	130252779	0.832423	0.832423	1
3296 rows × 7 columns
```



# Reflection
----

## What skills have you demonstrated in completing this notebook?

Your answer:

## What caused you the most difficulty?

Your answer: